### PR TITLE
Consul Upstream Load Balancing lab

### PR DESCRIPTION
**Pull request by danielehc**
Lab for load balancing tutorial being on PR: https://github.com/hashicorp/learn/pull/2381
 
Test lab at: https://www.katacoda.com/danielehc/scenarios/consul-envoy-load-balancing

### CODE REVIEW

**Review by freddygv — CHANGES_REQUESTED** [review]
This is looking good @danielehc, nice work!

I made some requests below, the most important one is about adjusting the default.

File: consul-upstream-load-balancing/provision.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+```
+- Complete! Move on to the next step.
+```
+
+Once this message appears, you are ready to continue.
+
+## Architecture for the lab
+
+![Consul service mesh load balancing with client service](./assets/consul-lb-envoy-client-service.png)
```
[the remark this answers is not in the document]
Comment (danielehc) [reply consul-upstream-load-balancing/provision.md:33]: Thanks. I was redoing some for the Workshop and got creative, but yes I think it is helpful to provide a visualization of the scenario.

File: consul-upstream-load-balancing/provision.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+### Configuration files
+
+While you wait for the environment to complete deploy you can review the files you are going to use for this lab later:
+
+| File                              | Description |
+|-----------------------------------|-------------|
+| `server.hcl`              | Server configuration file |
+| `agent.hcl`               | Client configuration file|
```
Comment (danielehc) [inline consul-upstream-load-balancing/provision.md:42]:
```suggestion
| `server.hcl`              | Server agent configuration file |
| `agent.hcl`               | Client agent configuration file|
```

File: consul-upstream-load-balancing/step1.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+This can be also verified using `consul members` on one of the containers
+
+`docker exec server consul members`{{execute}}
+
+```
+Node        Address          Status  Type    Build     Protocol  DC   Segment
+server-1    172.18.0.2:8301  alive   server  1.9.0dev  2         dc1  <all>
+client-1    172.18.0.3:8301  alive   client  1.9.0dev  2         dc1  <default>
+ingress-gw  172.18.0.6:8301  alive   client  1.9.0dev  2         dc1  <default>
+service-1   172.18.0.4:8301  alive   client  1.9.0dev  2         dc1  <default>
```
[the remark this answers is not in the document]
Comment (danielehc) [reply consul-upstream-load-balancing/step1.md:26]: I was trying to have a standard naming for docker containers that I could re-use in future scenarios, client in this case was both the name of the app/service but also showing how to get access to a service in the mesh using your machine.

File: consul-upstream-load-balancing/step2.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+Example output:
+
+```
+Config entry written: service-defaults/backend
+```
+
+### Configure a sticky session for service resolution
+
+A common requirements for many applications is to have the possibility to redirect all the requests from a specific client to the same server.
```
[the remark this answers is not in the document]
Comment (danielehc) [reply consul-upstream-load-balancing/step2.md:25]: For the first point that seems a good feature to have, I mean better reconstruct the "state" than deny requests. 
Do we have some more logic on that? I mean my understanding is that the extra component (in this case the header) was used to generate one hash that would remain consistent. 

I might need help in phrasing this but I think we can pass the message that "sticky session is consistent (reliable/predictable) given a stable service configuration, if the number of services we load balance on changes, some sessions might be routed to a different host after the change but will then maintain the new association."

File: consul-upstream-load-balancing/step3.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
+The default load balancing policy, _random_, is usually the best approach in scenarios where requests are homogeneous and the system is over-provisioned.
```
[the remark this answers is not in the document]
Comment (danielehc) [reply consul-upstream-load-balancing/step3.md:1]: Given the fact we have checks and that LB is made on an healthy pool the statement "best approach" is still valid if applied to round robin.

> The default load balancing policy, *round_robin*, is usually the best approach in scenarios where service nodes capacity is homogeneous and when you can rely on a health check policy that prevents failed nodes to be used for the scheduling. If you configure health checks for your services you'll automatically benefit from them during the load balancing process.

File: consul-upstream-load-balancing/finish.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
+## Review
+
+In this hands-on lab, you will operated a Consul datacenter, deployed using Docker, to configure load balancing policies on Envoy sidecar proxies.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/finish.md:3]:
```suggestion
In this hands-on lab, you operated a Consul datacenter that was deployed using Docker and configured load balancing policies on the service sidecar proxies.
```

File: consul-upstream-load-balancing/finish.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+Specifically, you:
+- Inspected the environment
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/finish.md:6]:
```suggestion
- Inspected the environment with the Consul CLI
```
Is that accurate?

File: consul-upstream-load-balancing/finish.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+- Verified the default *round_robin* load balancing policy
+- Used centralized configuration to set service defaults
+- Used the _maglev_ policy to configure load balancing with sticky session
+- Verified the new configuration
+- Uses the *least_request* load balancing policy
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/finish.md:11]:
```suggestion
- Used the *least_request* load balancing policy
```

File: consul-upstream-load-balancing/index.json [hunk 9]
```diff
@@ -0,0 +1,78 @@
+{
+    "pathwayTitle": "Load balancing in Consul service mesh with Envoy",
+    "icon": "fa-consul",
+    "title": "Load balancing in Consul service mesh with Envoy",
+    "description": "Configure different load balancing policies to share load across services in your service mesh configuring Envoy automatically",
+    "details": {
+        "steps": [
+            {
+                "title": "Provisioning Lab Infrastructure",
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/index.json:9]:
```suggestion
                "title": "Lab Infrastructure Provisioning",
```

File: consul-upstream-load-balancing/intro.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
+# Load balancing in Consul service mesh with Envoy
+
+In this hands-on lab, you will operate a Consul datacenter, deployed using Docker, to configure load balancing policies on Envoy sidecar proxies.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/intro.md:3]:
```suggestion
In this hands-on lab, you will operate a Consul datacenter deployed using Docker and configure load balancing policies on service sidecar proxies.
```

File: consul-upstream-load-balancing/intro.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+Specifically, you will:
+- Inspect the environment
+- Verify default *least_request* load balancing policy
+- Use centralized configuration to set service defaults
+- Used the _maglev_ policy to configure load balancing with sticky session
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/intro.md:9]:
```suggestion
- Use the _maglev_ policy to configure load balancing with sticky session
```

File: consul-upstream-load-balancing/intro.md [hunk 12]
```diff
@@ -0,0 +1,19 @@
+# Load balancing in Consul service mesh with Envoy
+
+In this hands-on lab, you will operate a Consul datacenter, deployed using Docker, to configure load balancing policies on Envoy sidecar proxies.
+
+Specifically, you will:
+- Inspect the environment
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/intro.md:6]:
```suggestion
- Inspect the environment with the Consul CLI
```

File: consul-upstream-load-balancing/step1.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (freddygv) [reply consul-upstream-load-balancing/step1.md:26]: I see, sounds good.

File: consul-upstream-load-balancing/step1.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
+You can check the environment using `docker ps`:
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step1.md:1]:
```suggestion
Before updating the load balancing policy, you can check that the environment has five Consul agents, two backend services, and that the load balancing policy is the default `round_robin`.  

First, use `docker ps`:
```

File: consul-upstream-load-balancing/index.json [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+                "text": "provision.md",
+                "courseData": "provision.courseData.sh",
+                "code": "provision.code.sh",
+                "verify": "provision.verify.sh"
+            },
+            {
+                "title": "Test environment and default load balancing policy",
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/index.json:16]:
```suggestion
                "title": "Verify the environment and load balancing policy",
```

File: consul-upstream-load-balancing/step2.md [hunk 15]
```diff
@@ -0,0 +1,111 @@
+Once verified you can access the _backend_ service and that the *round_robin* policy is applied you can apply new policies for the load balancing and verify how these affect the requests' resolution.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step2.md:1]:
```suggestion
Once verified, you can access the _backend_ service and that the *round_robin* policy is applied you can apply new policies for the load balancing and verify how these affect the requests' resolution.
```

File: consul-upstream-load-balancing/step2.md [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+
+### Configure service defaults
+
+In order to enable service resolution and apply load balancer policies you need to define the service protocol in a `service-defaults` configuration entry.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step2.md:5]:
```suggestion
In order to enable service resolution and apply load balancer policies, you need to define the service protocol in a `service-defaults` configuration entry.
```

File: consul-upstream-load-balancing/step2.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (freddygv) [reply consul-upstream-load-balancing/step2.md:25]: that's mostly accurate I would make a couple edits:

> sticky sessions are consistent (reliable/predictable) given a stable service configuration. If the number of backend hosts changes, a fraction of the sessions will be routed to a new host after the change.

If the number of backend services changes, the routing *will* change for some services.

File: consul-upstream-load-balancing/step3.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (freddygv) [reply consul-upstream-load-balancing/step3.md:1]: Hmm cool, then we just need to replace the word `random`

File: consul-upstream-load-balancing/step2.md [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+
+Example output:
+
+```
+Config entry written: service-resolver/backend
+```
+
+### Verify the policy is applied
+
+Once the policy is in place you can test it using the `curl` command and applying the `x-user-id` header to the request:
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step2.md:59]:
```suggestion
Once the policy is in place, you can test it using the `curl` command and applying the `x-user-id` header to the request:
```

File: consul-upstream-load-balancing/step2.md [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+  ],
+  "start_time": "2020-10-01T16:15:47.950151",
+  "end_time": "2020-10-01T16:15:47.950581",
+  "duration": "430.088µs",
+  "body": "Hello World",
+  "code": 200
+}
+```
+
+If you execute the previous command multiple times you will always be redirected to the same instance of the _backend_ service.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step2.md:81]:
```suggestion
Execute the curl command multiple times, you will always be redirected to the same instance of the _backend_ service.

`docker exec client curl -s localhost:9192 -H "x-user-id: 12345"`{{execute}}
```

File: consul-upstream-load-balancing/step2.md [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+
+### Check configuration
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step2.md:83]: This is great!

File: consul-upstream-load-balancing/step3.md [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+
+In scenarios where the different instances might be subject to substantial differences in terms of workload there are better approaches.
+
+Using the *least_request* policy permits Envoy to resolve requests using information on instance load level and select the one with the lowest load.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step3.md:5]:
```suggestion
Using the *least_request* policy permits Envoy sidecar proxies to resolve requests using information on instance load level and select the one with the lowest load.
```

File: consul-upstream-load-balancing/step3.md [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+Name           = "backend"
+LoadBalancer = {
+  Policy = "least_request"
+  LeastRequestConfig = {
+    ChoiceCount = "2"
+  }
+}
+```
+
+This configuration creates a `service-resolver` configuration, for the service `backend` that for every request picks 2 (as expressed by `ChoiceCount`) random instances of the service and resolves to the one with the least amount of load.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step3.md:22]:
```suggestion
This configuration creates a `service-resolver` load balancing policy, for every request to the `backend` service 2 (as expressed by `ChoiceCount`) random instances of that service are selected and the traffic is routed to the one with the least amount of load.
```

File: consul-upstream-load-balancing/step3.md [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+
+Example output:
+
+```
+Config entry written: service-resolver/backend
+```
+
+### Verify the policy is applied
+
+Once the policy is in place you can test it using the `curl` command and applying the `x-user-id` header to the request:
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step3.md:36]:
```suggestion
Once the policy is in place, you can test it using the `curl` command and applying the `x-user-id` header to the request:
```

File: consul-upstream-load-balancing/step3.md [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+  ],
+  "start_time": "2020-10-01T16:25:47.950151",
+  "end_time": "2020-10-01T16:25:47.950581",
+  "duration": "420.066µs",
+  "body": "Hello World",
+  "code": 200
+}
+```
+
+Executing the command multiple times you can verify that, despite you are still applying the header that was used by the previous policy the request gets served by different instances of the service.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step3.md:58]:
```suggestion
Execute the command multiple times to verify that, despite the user id in the header, the request gets served by different instances of the service.

`docker exec client curl -s localhost:9192 -H "x-user-id: 12345"`{{execute}}
```

File: consul-upstream-load-balancing/step4.md [hunk 24]
```diff
@@ -0,0 +1,47 @@
+
+The load balancing policy for the datacenter applies also to the service resolution performed by ingress gateways. So, once you configured the policies for the services and tested it internally using the client service, you can introduce an ingress gateway in your configuration and the same policies will be now respected by external requests being served by your Consul datacenter.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step4.md:2]:
```suggestion
The load balancing policy for the service sidecar proxies also applies to the service resolution performed by ingress gateways. Once you configured the policies for the services and tested it internally using the client service, you can introduce an ingress gateway in your configuration and the same policies will be now respected by external requests being served by your Consul datacenter.
```

File: consul-upstream-load-balancing/step4.md [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+`docker exec server consul config write /etc/consul.d/hash-resolver.hcl`{{execute}}
+
+Example output:
+
+```
+Config entry written: service-resolver/backend
+```
+
+### Verify the policy is applied
+
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step4.md:24]: Do you want to add a statement about refreshing the browser to test too?

File: consul-upstream-load-balancing/step4.md [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+  ],
+  "start_time": "2020-10-01T19:11:57.250151",
+  "end_time": "2020-10-01T19:11:57.250581",
+  "duration": "430.088µs",
+  "body": "Hello World",
+  "code": 200
+}
+```
+
+If you execute the previous command multiple times you will always be redirected to the same instance of the _backend_ service.
```
Comment (kaitlincarter-hc) [inline consul-upstream-load-balancing/step4.md:47]:
```suggestion
Execute the curl command multiple times you, will always be redirected to the same instance of the _backend_ service.

`curl -s backend.ingress.consul:8080 -H "x-user-id: 12345"`{{execute}}
```

**Review by kaitlincarter-hc — APPROVED** [review]
This is a really great lab! I'm excited to get is published next week.

File: consul-upstream-load-balancing/step4.md (hunk 25 above)
Comment (danielehc) [reply consul-upstream-load-balancing/step4.md:24]: Browser cannot pass the header so easily unless we ask them to use smt like poastman

File: consul-upstream-load-balancing/step3.md [hunk 27]
```diff
@@ -0,0 +1,67 @@
+The default load balancing policy, _random_, is usually the best approach in scenarios where requests are homogeneous and the system is over-provisioned.
```
Comment (danielehc) [inline consul-upstream-load-balancing/step3.md:1]:
```suggestion
The default load balancing policy, *round_robin*, is usually the best approach in scenarios where requests are homogeneous and the system is over-provisioned.
```

File: consul-upstream-load-balancing/step2.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (danielehc) [reply consul-upstream-load-balancing/step2.md:25]: I am adding this later on in the

File: consul-upstream-load-balancing/step2.md [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+  "duration": "430.088µs",
+  "body": "Hello World",
+  "code": 200
+}
+```
+
+Execute the curl command multiple times, you will always be redirected to the same instance of the _backend_ service.
+
+`docker exec client curl -s localhost:9192 -H "x-user-id: 12345"`{{execute}}
+
```
Comment (danielehc) [inline consul-upstream-load-balancing/step2.md:84]:
```suggestion

<div style="background-color:#fcf6ea; color:#866d42; border:1px solid #f8ebcf; padding:1em; border-radius:3px; margin:24px 0;">
  <p><strong>Note:</strong> sticky sessions are consistent given a stable service configuration. If the number of backend hosts changes, a fraction of the sessions will be routed to a new host after the change.
</p></div>
```

**Pull request merged**